src/server/worker/spawner/service.go

```diff
@@ -1,30 +1,69 @@
 package spawner
 
-func (s *Spawner) RunService(pachClient *client.APIClient) error {
+type serviceItem struct {
+	serviceCtx context.Context
+	commitInfo *pfs.CommitInfo
+}
+
+// Runs the given callback with the latest commit for the pipeline.  The given
+// context will be canceled if a newer commit is ready.
+func forLatestCommit(
+	pachClient *client.APIClient,
+	pipelineInfo *pps.PipelineInfo,
+	cb func(context.Context, *pfs.CommitInfo) error,
+) error {
 	ctx := pachClient.Ctx()
-	commitIter, err := pachClient.SubscribeCommit(a.pipelineInfo.Pipeline.Name, a.pipelineInfo.OutputBranch, "", pfs.CommitState_READY)
+
+	commitIter, err := pachClient.SubscribeCommit(pipelineInfo.Pipeline.Name, pipelineInfo.OutputBranch, "", pfs.CommitState_READY)
 	if err != nil {
 		return err
 	}
-	defer commitIter.Close()
+
 	var serviceCtx context.Context
 	var serviceCancel func()
-	var dir string
+
+	itemChan := make(chan serviceItem)
+
+	cancel := func() {
+		close(itemChan)
+		if serviceCancel != nil {
+			serviceCancel()
+		}
+	}
+
+	defer cancel()
+
+	// Run the callback in a single goroutine so we can ensure it runs in order
+	// and non-concurrently with itself.
+	go func() {
+		for item := range itemChan {
+			cb(item.ctx, item.commitInfo)
+		}
+	}
+
 	for {
-		commitInfo, err := commitIter.Next()
-		if err != nil {
+		if commitInfo, err := commitIter.Next(); err != nil {
 			return err
+		} else if commitInfo.Finished == nil {
+			cancel()
+			serviceCtx, serviceCancel = context.WithCancel(ctx)
+			itemChan <- serviceItem{serviceCtx, commitInfo}
```

Comment:
I think these channel operations all need to be in selects that also select on the context being cancelled, otherwise they can get caught in a deadlock. However, I think you can get rid of the channel completely if you use `SubscribeCommitF` and put all of the logic in the callback.

Reply:
I don't think I can use `SubscribeCommitF` here, because it specifically needs to cancel the running service if a new commit comes in.  That said, the code here is totally wrong and needs some work.

Reply:
I'm not quite sure why that would prevent doing this with `SubscribeCommitF` I was thinking the code would look something like this:

```golang
var ctx context.Context
var cancel func()
return pachClient.SubscribeCommitF(pipelineInfo.Pipeline.Name, pipelineInfo.OutputBranch, "", pfs.CommitState_READY, func(ci *pfs.CommitInfo) error {
  if cancel != nil {
    cancel()
    ctx, cancel = context.WithCancel(pachClient.Ctx())
    go cb(ctx, ci)
    return nil
  }
})
```

I'm probably missing why this won't work though.

Reply:
Hmm, thanks - that looks pretty close, although I'd like to ensure that the old callback has finished before starting the next one, which will complicate it a little more.

Reply:
This should be fixed - mostly used your suggestion with an added `errgroup.Group` to wait for the service to return before starting the next one.